lib/raygun.dart

```diff
@@ -0,0 +1,191 @@
+
+enum ReactionState {
+  add,
+  remove,
+}
+
+enum PinState {
+  pin,
+  unpin,
+}
+
+enum EmbedState {
+  enabled,
+  disabled,
+}
+
+class Message {
+  late String id;
+  late String conversationId;
+  late String senderId;
+  late DateTime date;
+  late bool pinned;
+  late List<Reaction> reactions;
+  late String replied;
+  late List<String> value;
+  late Map<String, String> metadata;
+}
+
+class Reaction {
+  late String emoji;
+  late List<String> senderId;
+}
+
+class SenderId {
+  late Uid uid;
+}
+
+class Raygun {
+  Pointer<G_RayGunAdapter> pRaygun;
+  Raygun(this.pRaygun);
+
+  List<Message> getMessages(String conversationID) {
+    G_FFIResult_FFIVec_Message result = bindings.raygun_get_messages(
+        pRaygun, conversationID.toNativeUtf8().cast<Int8>());
+
+    if (result.error.address.toString() != "0") {
+      throw WarpException(result.error);
+    }
+
+    List<Message> msgs = [];
+    // Iterate over the messages
+    int length = result.data.ref.len;
+    for (int i = 0; i < length; i++) {
+      Message message = Message();
+      Pointer<G_Message> pMsg = result.data.ref.ptr.elementAt(i).value;
+
+      // Message ID
+      message.id = bindings.message_id(pMsg).value.toString();
+      // Conversation ID
+      message.conversationId =
+          bindings.message_conversation_id(pMsg).value.toString();
+      // Sender ID, DID only
+      Pointer<G_SenderId> senderIdPointer = bindings.message_sender_id(pMsg);
+      Pointer<G_DID> senderDid =
+          bindings.sender_id_get_did_key(senderIdPointer);
+      message.senderId = DID(senderDid).toString();
+      // DateTime
+      // TODO: test required
+      // The value from Rust is integer (UTC from Chrono ).
+      // Possibly DateTime.fromMillisecondsSinceEpoch() can be used.
+      message.date = DateTime(bindings.message_date(pMsg).value);
+      // Reactions
+      List<Reaction> reactions = [];
+      Pointer<G_FFIVec_Reaction> pReactions = bindings.message_reactions(pMsg);
+      int reactionsLen = pReactions.ref.len;
+      for (int j = 0; j < reactionsLen; j++) {
+        Reaction reaction = Reaction();
+        Pointer<G_Reaction> pReaction = pReactions.ref.ptr.elementAt(j).value;
+        reaction.emoji = bindings.reaction_emoji(pReaction).toString();
+        Pointer<G_FFIVec_SenderId> reactionSendersId =
+            bindings.reaction_users(pReaction);
+        int reactionSendersIdLen = reactionSendersId.ref.len;
+        for (int k = 0; k < reactionSendersIdLen; k++) {
+          reaction.senderId
+              .add(reactionSendersId.ref.ptr.elementAt(k).value.toString());
+        }
+      }
+      message.reactions = reactions;
+      // Pinned
+      message.pinned = bindings.message_pinned(pMsg) == 0 ? false : true;
+      // Message body
+      Pointer<G_FFIVec_String> lines = bindings.message_lines(pMsg);
+      int lineLen = lines.ref.len;
+      for (int j = 0; j < lineLen; j++) {
+        message.value.add(lines.ref.ptr.value.toString());
```

Comment:
From glancing over it, it looks good. Dont forget to check to make sure the pointers arent a nullptr before adding converting them to a string and freeing the pointer once done. Also I thought it used something along the lines of `.cast<Utf8>().toDartString()` when converting it to a dart string?

Reply:
1. I will update for the pointer checking/freeing part.
2. For **toDartString**, it yields an error if used for the value: 
- The method 'toDartString' isn't defined for the type 'Pointer'. Try correcting the name to the name of an existing method, or defining a method named 'toDartString'

Reply:
For the first part (checking null pointer) it seems not available since the type is G_FFIVec_String (we need G_FFIResult_* to access to the error, isn't it?). 
For the toDartString, I updated the line of adding (also found my mistake from earlier commits).